---
title: "RNA Seq Part 3"
output: RNA_seq_p3
---

*Run* button within the chunk or by pressing *Cmd+Shift+Enter*. 

*Insert Chunk* button on the toolbar or by pressing *Cmd+Option+I*.


```{r}
library(dplyr)
library(Seurat)
library(patchwork)

# Load the PBMC dataset
pbmc.data <- Read10X(data.dir = "~/filtered_gene_bc_matrices/hg19/")
# Initialize the Seurat object with the raw (non-normalized data).
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc
```

```{r}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")
```

```{r}
# Visualize QC metrics as a violin plot
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

```{r}
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2
```

#normalizes gene expression for each cell to minimize sequencing depth differences
```{r}
pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", scale.factor = 10000)
```
```{r}
pbmc <- NormalizeData(pbmc)
```

```{r}
pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(pbmc), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(pbmc)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 + plot2
```

#standardizes gene expression for all genes
```{r}
all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes)
```

#creates and assigns the PCA to a variables
```{r}
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))
```

#prints the first five principal components for the PCA
```{r}
print(pbmc[["pca"]], dims = 1:5, nfeatures = 5)
```

#visualizes the PCA structure based on the first two principal components
```{r}
VizDimLoadings(pbmc, dims = 1:2, reduction = "pca")
```

#plots cells in PCA plot with each cluster being a different color
```{r}
DimPlot(pbmc, reduction = "pca") + NoLegend()
```

#creates a heat map for the first principal component
```{r}
DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)
```

#creates a heat map for the first 15 principal components 
```{r}
DimHeatmap(pbmc, dims = 1:15, cells = 500, balanced = TRUE)
```

#plots the standard deviation for each principal component
```{r}
ElbowPlot(pbmc)
```

#constructs SNN graph and performs community detection to define clusters
```{r}
pbmc <- FindNeighbors(pbmc, dims = 1:10)
pbmc <- FindClusters(pbmc, resolution = 0.5)
```
#gives teh first 5 samples
```{r}
head(Idents(pbmc), 5)
```

#creates and assigns UMAP based on first 10 principal components
```{r}
pbmc <- RunUMAP(pbmc, dims = 1:10)
```
#displays the generated umap
```{r}
DimPlot(pbmc, reduction = "umap")
```

#saves the Seurat object as an .rds file
```{r}
saveRDS(pbmc, file = "pbmc_tutorial.rds")
```

```{r}
# find all markers of cluster 2
cluster2.markers <- FindMarkers(pbmc, ident.1 = 2)
head(cluster2.markers, n = 5)
```

```{r}
# find all markers distinguishing cluster 5 from clusters 0 and 3
cluster5.markers <- FindMarkers(pbmc, ident.1 = 5, ident.2 = c(0, 3))
head(cluster5.markers, n = 5)
```

```{r}
# find markers for every cluster compared to all remaining cells, report only the positive
# ones
pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE)
pbmc.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1)
```

#uses ROC test to identify markers uniquely upregulated to cluster 0
```{r}
cluster0.markers <- FindMarkers(pbmc, ident.1 = 0, logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
```

#generates a VinPlot for MSA4A1 and CD79A
```{r}
VlnPlot(pbmc, features = c("MS4A1", "CD79A"))
```

```{r}
# you can plot raw counts as well
VlnPlot(pbmc, features = c("NKG7", "PF4"), slot = "counts", log = TRUE)
```

#visualizes the gene expression patterns across the clusters 
```{r}
FeaturePlot(pbmc, features = c("MS4A1", "GNLY", "CD3E", "CD14", "FCER1A", "FCGR3A", "LYZ", "PPBP",
    "CD8A"))
```

#plots a heatmap for the top 10 marker genes per cluster
```{r}
pbmc.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1) %>%
    slice_head(n = 10) %>%
    ungroup() -> top10
DoHeatmap(pbmc, features = top10$gene) + NoLegend()
```

#assigns human-readable cell type to clusters then replots the data
```{r}
new.cluster.ids <- c("Naive CD4 T", "CD14+ Mono", "Memory CD4 T", "B", "CD8 T", "FCGR3A+ Mono",
    "NK", "DC", "Platelet")
names(new.cluster.ids) <- levels(pbmc)
pbmc <- RenameIdents(pbmc, new.cluster.ids)
DimPlot(pbmc, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()
```

#loads ggplot2 to create then save the UMAP plot as a JPEG
```{r}
library(ggplot2)
plot <- DimPlot(pbmc, reduction = "umap", label = TRUE, label.size = 4.5) + xlab("UMAP 1") + ylab("UMAP 2") +
    theme(axis.title = element_text(size = 18), legend.text = element_text(size = 18)) + guides(colour = guide_legend(override.aes = list(size = 10)))
ggsave(filename = "pbmc3k_umap.jpg", height = 7, width = 12, plot = plot, quality = 50)
```

#saves the final Seurat object
```{r}
saveRDS(pbmc, file = "pbmc3k_final.rds")
```

